apps/system/js/advanced_telemetry.js

```diff
@@ -324,6 +362,84 @@
     }
   };
 
+  // Check if there are existing metrics that need to be merged, if so,
+  // merge them and then transmit.
+  AT.prototype.mergeSend = function mergeSend(payloadNew, send) {
+    var self = this;
+    // Check if there are metrics in asyncStorage that need to be merged.
+    // If not, then we can just send them.
+    asyncStorage.getItem(AT.METRICS_KEY, function(value) {
+      if (value) {
+        mergeMetrics(value, payloadNew, self);
+      } else {
+        if (send) {
+          self.transmit(payloadNew);
+        } else {
+          asyncStorage.setItem(AT.METRICS_KEY, payloadNew);
+        }
+      }
+    });
+
+    // Merge the new metrics into the old metrics and then transmit the
+    // result to the server.
+    function mergeMetrics(payloadOld, payloadNew, self) {
+      mergeKeyed(payloadOld.payload.keyedHistograms, payloadOld, payloadNew);
+      mergeLists(payloadOld.payload.addonHistograms,
+                 payloadNew.payload.addonHistograms);
+      if (send) {
+        self.transmit(payloadOld);
+        asyncStorage.setItem(AT.METRICS_KEY, null);
+      } else {
+        asyncStorage.setItem(AT.METRICS_KEY, payloadOld);
+      }
+    }
+  };
+
+  function mergeHistogram(oldHist, newHist) {
+    oldHist.sum += newHist.sum;
+    oldHist.log_sum += newHist.log_sum;
+    oldHist.log_sum_squares += newHist.log_sum_squares;
+
+    for(var i = 0; i < oldHist.counts.length; i++) {
+      oldHist.counts[i] += newHist.counts[i];
+    }
+  }
+
+  function mergeLists(oldHistList, newHistList) {
+    var histMap = new Map();
+
+    for(var keyOld in oldHistList) {
+      histMap.set(keyOld, oldHistList[keyOld]);
+    }
+
+    for(var keyNew in newHistList) {
+      if(histMap.has(keyNew)) {
+        mergeHistogram(oldHistList[keyNew], newHistList[keyNew]);
+      } else {
+        oldHistList[keyNew] = newHistList[keyNew];
+      }
+    }
+  }
+
+  function mergeKeyed(keyedOld, payloadOld, payloadNew) {
+    var addonMapKey = new Map();
+
+    for(var keyOld in keyedOld) {
+      addonMapKey.set(keyOld, keyedOld[keyOld]);
+    }
+
+    var keyedNew = payloadNew.payload.keyedHistograms;
```

Comment:
Having variables for the old a new keyed histograms would increase readability (var oldKeyed = payloadOld.payload.keyedHistograms; var newKeyed = payloadNew.payload.keyedHistograms)